crates/embucket-functions/src/encryption/decrypt_raw.rs

```diff
@@ -0,0 +1,346 @@
+///
+/// Decrypts ciphertext produced by `ENCRYPT_RAW` using AES-GCM.
+///
+/// Syntax: `DECRYPT_RAW(<ciphertext>, <key>, <iv> [, <aad>, <method>, <tag>])`
+///
+/// Arguments:
+/// - `ciphertext`: encrypted bytes.
+/// - `key`: encryption key (16, 24 or 32 bytes).
+/// - `iv`: initialization vector.
+/// - `aad`: optional additional authenticated data.
+/// - `method`: encryption method, only `'AES-GCM'` supported.
+/// - `tag`: authentication tag.
+///
+/// Returns: decrypted binary data.
+#[derive(Debug)]
+pub struct DecryptRawFunc {
+    signature: Signature,
+}
+
+impl Default for DecryptRawFunc {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl DecryptRawFunc {
+    #[must_use]
+    pub fn new() -> Self {
+        Self {
+            signature: Signature::user_defined(Volatility::Immutable),
+        }
+    }
+}
+
+impl ScalarUDFImpl for DecryptRawFunc {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn name(&self) -> &'static str {
+        "decrypt_raw"
+    }
+
+    fn signature(&self) -> &Signature {
+        &self.signature
+    }
+
+    fn return_type(&self, _arg_types: &[DataType]) -> DFResult<DataType> {
+        Ok(DataType::Binary)
+    }
+
+    #[allow(clippy::too_many_lines)]
+    fn coerce_types(&self, arg_types: &[DataType]) -> DFResult<Vec<DataType>> {
+        // Handle NULL values by coercing them to the expected types
+        let coerced_types: Vec<DataType> = arg_types
+            .iter()
+            .enumerate()
+            .map(|(i, arg_type)| {
+                if arg_type.is_null() {
+                    match i {
+                        0 | 1 | 2 | 3 | 5 => DataType::Binary, // ciphertext, key, iv, aad, tag
+                        4 => DataType::Utf8,                   // method
+                        _ => arg_type.clone(),
+                    }
+                } else {
+                    arg_type.clone()
+                }
+            })
+            .collect();
+
+        // Validate the coerced signature matches one of our expected signatures
+        let format_types = |types: &[DataType]| -> String {
+            format!(
+                "({})",
+                types
+                    .iter()
+                    .map(|t| format!("{t:?}"))
+                    .collect::<Vec<_>>()
+                    .join(", ")
+            )
+        };
+
+        match coerced_types.len() {
+            3 => {
+                if matches!(coerced_types[0], DataType::Binary)
+                    && matches!(coerced_types[1], DataType::Binary)
+                    && matches!(coerced_types[2], DataType::Binary)
+                {
+                    Ok(coerced_types)
+                } else {
+                    Err(InvalidArgumentTypesSnafu {
+                        function_name: self.name().to_uppercase(),
+                        types: format_types(arg_types),
+                    }
+                    .build()
+                    .into())
```

Comment:
Here and in other PR lines, when returning error 
it's possible to use `.fail()` to create errored `Result<>`
and then `?` immediately raises error

```suggestion
                    InvalidArgumentTypesSnafu {
                        function_name: self.name().to_uppercase(),
                        types: format_types(arg_types),
                    }.fail()?
```